### PR TITLE
Add Opentelemetry support for Logs / Traces

### PR DESCRIPTION
**Pull request by mhelleborg**
## Summary

Adds support for publishing OpenTelemetry traces and logs via OTLP. Logs are correlated with traces with both spanId & traceId.

## Configuration

Requires `DOLITTLE__RUNTIME__OPENTELEMETRY__ENDPOINT` to be set, and supports disabling logs or traces individually.

```csharp
[Configuration("opentelemetry")]
public class OpenTelemetryConfiguration
{
    /// <summary>
    /// OTLP reporting endpoint
    /// </summary>
    public string Endpoint { get; set; }
    
    public string ServiceName { get; set; } = "dolittle-runtime";
    
    public bool Logging { get; set; } = true;
    public bool Tracing { get; set; } = true;
}
```

### CODE REVIEW

File: Source/Server/Server.csproj [hunk 1]
```diff
@@ -16,6 +16,13 @@
         <PackageReference Include="Dolittle.Contracts" Version="$(ContractsVersion)" />
         <PackageReference Include="Microsoft.Extensions.Hosting" Version="$(MicrosoftExtensionsVersion)" />
         <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="$(MicrosoftExtensionsVersion)" />
+        <PackageReference Include="MongoDB.Driver.Core.Extensions.OpenTelemetry" Version="1.0.0" />
```
Comment (woksin) [inline Source/Server/Server.csproj:19]: Same comment as on the DatabaseConnection.cs file
```suggestion
        <PackageReference Include="MongoDB.Driver.Core.Extensions.OpenTelemetry" Version="$(MongoDBOpenTelemetryVersion)" />
```

File: Source/Actors/Hosting/HostBuilderExtensions.cs [hunk 2]
```diff
@@ -36,5 +38,6 @@ public static IHostBuilder AddActorSystem(this IHostBuilder builder)
                         new TestProvider(new TestProviderOptions(), new InMemAgent()),
                         new PartitionIdentityLookup())
                     .WithDiscoveredClusterKinds(provider)))
+            .AddSingleton(provider => provider.GetRequiredService<ActorSystem>().Root.WithTracing())
```
[the remark this answers is not in the document]
Comment (woksin) [reply Source/Actors/Hosting/HostBuilderExtensions.cs:41]: As I commented below, if `IRootContext` is supposed to be added to the IoC then just disregard this comment :)

File: Source/Events.Store/EventStore.cs [hunk 3]
```diff
@@ -20,18 +21,22 @@ public class EventStore : IEventStore
...
 
     /// <summary>
     /// Initializes a new instance of the <see cref="EventStore"/> class.
     /// </summary>
     /// <param name="getEventStoreClient">The factory to use to get the <see cref="EventStoreActor"/> client.</param>
     /// <param name="logger">The logger to use for logging.</param>
-    public EventStore(Func<TenantId, EventStoreClient> getEventStoreClient, ILogger logger, Func<TenantId, IFetchCommittedEvents> getCommittedEventsFetcher)
+    /// <param name="getCommittedEventsFetcher"></param>
+    /// <param name="rootContext">Proto root context. Allows middleware to be used</param>
+    public EventStore(Func<TenantId, EventStoreClient> getEventStoreClient, ILogger logger, Func<TenantId, IFetchCommittedEvents> getCommittedEventsFetcher, IRootContext rootContext)
```
[the remark this answers is not in the document]
Comment (mhelleborg) [reply Source/Events.Store/EventStore.cs:34]: Yes, it's to enable us to use it wrapped with middleware. The one on ActorSystem does not include middleware

File: Source/Events.Store.MongoDB/DatabaseConnection.cs [hunk 4]
```diff
@@ -43,6 +44,7 @@ public DatabaseConnection(IOptions<EventStoreConfiguration> configuration, IConf
             Servers = config.Servers.Select(MongoServerAddress.Parse),
             GuidRepresentation = GuidRepresentation.Standard,
             MaxConnectionPoolSize = config.MaxConnectionPoolSize,
+            ClusterConfigurator = cb => cb.Subscribe(new DiagnosticsActivityEventSubscriber())
```
[the remark this answers is not in the document]
Comment (mhelleborg) [reply Source/Events.Store.MongoDB/DatabaseConnection.cs:47]: Done :)

File: Source/Server/Server.csproj [hunk 5]
```diff
@@ -16,6 +16,13 @@
...
+        <PackageReference Include="OpenTelemetry.Exporter.OpenTelemetryProtocol" Version="1.3.0" />
+        <PackageReference Include="OpenTelemetry.Exporter.OpenTelemetryProtocol.Logs" Version="1.3.0-rc.2" />
+        <PackageReference Include="OpenTelemetry.Extensions.Hosting" Version="1.0.0-rc9.4" />
+        <PackageReference Include="OpenTelemetry.Instrumentation.AspNetCore" Version="1.0.0-rc9.4" />
+        <PackageReference Include="OpenTelemetry.Instrumentation.GrpcNetClient" Version="1.0.0-rc9.4" />
+        <PackageReference Include="OpenTelemetry.Instrumentation.Http" Version="1.0.0-rc9.4" />
```
[the remark this answers is not in the document]
Comment (mhelleborg) [reply Source/Server/Server.csproj:25]: Will clean this up when the OT versions are unified and released

**Review by woksin — APPROVED** [review]

**Pull request merged**